crates/sui-framework/packages/bridge/sources/limiter.move

```diff
@@ -0,0 +1,255 @@
+        // USD notional value, 0 DP accuracy
+        notional_values: VecMap<TypeName, u64>,
+        // Per hour transfer amount for each bridge route
+        transfer_records: VecMap<BridgeRoute, TransferRecord>,
+    }
+
+    struct TransferRecord has store {
+        hour_head: u64,
+        hour_tail: u64,
+        per_hour_amounts: vector<u64>,
+        // total amount in USD, 0 DP accuracy
+        total_amount: u64
+    }
+
+    public fun new(): TransferLimiter {
+        TransferLimiter {
+            transfer_limits: vec_map::empty(),
+            notional_values: vec_map::empty(),
+            transfer_records: vec_map::empty()
+        }
+    }
+
+    // Current hour since unix epoch
+    fun current_hour_since_epoch(clock: &Clock): u64 {
+        clock::timestamp_ms(clock) / 3600000
+    }
+
+    public fun check_and_record_transfer<T>(
+        clock: & Clock,
+        self: &mut TransferLimiter,
+        route: BridgeRoute,
+        amount: u64
+    ) {
+        // Create record for route if not exists
+        if (!vec_map::contains(&self.transfer_records, &route)) {
+            vec_map::insert(&mut self.transfer_records, route, TransferRecord {
+                hour_head: 0,
+                hour_tail: 0,
+                per_hour_amounts: vector[],
+                total_amount: 0
+            })
+        };
+        let record = vec_map::get_mut(&mut self.transfer_records, &route);
+        let current_hour_since_epoch = current_hour_since_epoch(clock);
+
+        // First clean up old transfer histories
+        remove_stale_hour_maybe(record, current_hour_since_epoch);
+        // Backfill missing hours
+        append_new_empty_hours(record, current_hour_since_epoch);
+
+        // Get limit for the route
+        let route_limit = option::destroy_with_default(
+            vec_map::try_get(&self.transfer_limits, &route),
+            DEFAULT_TRANSFER_LIMIT
+        );
+
+        // Compute notional amount
+        let coin_type = type_name::get<T>();
+        let notional_amount = *vec_map::get(&self.notional_values, &coin_type) * amount;
+        let notional_amount = notional_amount / pow(10, treasury::token_decimals<T>());
+
+        // Check if transfer amount exceed limit
+        assert!(record.total_amount + notional_amount <= route_limit, ETransferAmountExceedLimit);
+
+        // Record transfer value
+        let new_amount = vector::pop_back(&mut record.per_hour_amounts) + notional_amount;
+        vector::push_back(&mut record.per_hour_amounts, new_amount);
+        record.total_amount = record.total_amount + notional_amount;
+    }
+
+    fun append_new_empty_hours(self: &mut TransferRecord, current_hour_since_epoch: u64) {
+        if (self.hour_head == current_hour_since_epoch) {
+            return; // nothing to backfill
+        };
+
+        // If tail is even older than 24 hours ago, advance it to that.
+        let target_tail = current_hour_since_epoch - 24;
```

Comment:
should it be 24 or 23? (same to line 128)

Reply:
should be 24? We want to keep full 24 hours records, 23 hours back + 1 current hour